### ISSUE TITLE
skewed generalized t

### ISSUE THREAD

**Issue by spinkney** [spinkney/helpful_stan_functions#12]
https://discourse.mc-stan.org/t/ideas-for-modeling-systematically-skewed-outliers/636/7

**Comment by spinkney** [comment]
```stan
functions{
  real sgt_lpdf(vector x, real mu, real s, real l, real p, real q)
  { // Skewed generalised t
    int N = dims(x)[1];
    real lz1 = lbeta(1.0 / p, q);
    real lz2 = lbeta(2.0 / p, q - 1.0 / p);
    real v = q^(-1.0/p) * inv_sqrt( ( 3 * square(l) + 1 ) * exp(lbeta(3.0/p, q - 2.0/p) - lz1) - 4 * square(l *exp(lz2-lz1)) );
    real m = 2 * v * s * l * q^(1.0/p) * exp(lz2 - lz1);
    real out1 = N * (log(p) - log(v) - log(s) - lmultiply(1/p, q) - lz1);
    real out2 = 0;
    
    for (n in 1:N) {
      real r = x[n] - mu + m;
      if (r < 0)
        out2 -= log1p((-r)^p / (q * (v * s * (1 - l))^p));
      else
        out2 -= log1p(r^p / (q * (v * s * (1 + l))^p));
    }
    return fma(out2, (1/p + q), out1);
  }
  
    real sgt_slow_lpdf(vector x, real mu, real s, real l, real p, real q) {
    // Skewed generalised t
    int N;
    real lz1;
    real lz2;
    real v;
    real m;
    real r;
    real out;
    N = dims(x)[1];
    lz1=lbeta(1.0/p,q);
    lz2=lbeta(2.0/p,q-1.0/p);
    v=q^(-1.0/p)*((3*l^2+1)*exp(lbeta(3.0/p,q-2.0/p)-lz1)-4*l^2*exp(lz2-lz1)^2)^(-0.5);
    m=2*v*s*l*q^(1.0/p)*exp(lz2-lz1);
    out=0;
    for (n in 1:N) {
      r=x[n]-mu+m;
      if (r<0)
      	     out=out+log(p)-log( 2*v*s*q^(1.0/p)*exp(lz1)* ( fabs(r)^p /(q*(v*s)^p*(l*(-1)+1)^p)+1 )^(1.0/p+q) );
      else
      	     out=out+log(p)-log(2*v*s*q^(1.0/p)*exp(lz1)*(fabs(r)^p /(q*(v*s)^p*(l*(1)+1)^p)+1)^(1.0/p+q));
    }
    return out;
  }
  
}
data {
  int<lower=0> N;
  vector[N] x;
  int flag;
}
parameters {
  real mu; 
  real<lower=0> sigma; 
  real<lower=-.99,upper=0> l; 
  real<lower=1,upper=10> p; 
  real<lower=3.0/p,upper=p*50> q; 
}
model {
  mu ~ normal(0,9.0/sqrt(N));
  l ~ normal(0,.5);
  p ~ lognormal(log(2),1);
  q ~ gamma(2,0.1);
  
  if(flag == 0)  x ~ sgt_slow(mu, sigma, l, p, q);
  else x ~ sgt(mu, sigma, l, p, q);
}
```

```r
library(cmdstanr)
library(sgt)

N <- 300
x <- rsgt(N, mu = 0, sigma = 1, lambda = 0, p = 2, q = Inf,
          mean.cent = TRUE, var.adj = TRUE)

sgt_mod <- cmdstan_model("../distributions/sgt.stan")

fit <- sgt_mod$sample(
  data = list(N = N,
              x = x,
              flag = 1),
  parallel_chains = 2,
  chains = 2,
  init = 1,
  seed = 12312,
  adapt_delta = 0.8,
  max_treedepth = 10,
  iter_sampling = 500,
  iter_warmup = 500
)
```

**Comment by spinkney** [comment]
This is slightly faster
```stan
 real sgt_lpdf(vector x, real mu, real s, real l, real p, real q)
  { // Skewed generalised t
    int N = num_elements(x);
    real z1 = beta(1.0 / p, q);
    real z2 = beta(2.0 / p, q - 1.0 / p);
    real v = q^(-1.0/p) * inv_sqrt( ( 3 * square(l) + 1 ) * beta(3.0/p, q - 2.0/p) / z1 - 4 * square(l * z2 / z1) ) ;
    real m = 2 * v * s * l * q^(1.0/p) * z2 / z1;
    real out1 = N * (log(p) - log(v) - log(s) - lmultiply(1/p, q) - lbeta(1.0 / p, q));
    real out2 = 0;
    
    for (n in 1:N) {
      real r = x[n] - mu + m;
      if (r < 0)
        out2 -= log1p((-r)^p / (q * (v * s * (1 - l))^p));
      else
        out2 -= log1p(r^p / (q * (v * s * (1 + l))^p));
    }
    return fma(out2, (1/p + q), out1);
  }
```

**Comment by spinkney** [comment]
closed with #45
